Load Solidus engine extension files automatically

aldesantis opened the pull request:
Improves `EngineExtensions::Decorators` to automatically load the right files when Solidus engines are installed. For instance, if solidus_backend is installed, it will tell Rails to load `lib/views/backend` and `lib/controllers/backend`, as well as the decorators in `lib/decorators/backend`.

This also renames `SolidusSupport::EngineExtensions::Decorators` to `SolidusSupport::EngineExtensions`, since the module doesn't only deal with decorators anymore.

aldesantis committed:
Load Solidus engine extension files automatically

By cleanly separating extension files for the different Solidus
engines, we can only load solidus_backend files when the engine
itself is available rather than doing it all the time.

In turn, this allows extensions to only actually require the parts
of Solidus that they really need.

kennyadsl commented on lib/solidus_support/engine_extensions/decorators.rb:
```diff
@@ -4,33 +4,62 @@ module SolidusSupport
-            Dir.glob(base_path.join('*')) do |decorators_folder|
+            # Add decorators folder to the Rails autoloader. This tells Zeitwerk to treat paths
+            # such as app/decorators/controllers as roots.
+            Dir.glob(decorators_root.join('*')) do |decorators_folder|
```
`decorators_root.glob('*')`

kennyadsl commented on lib/solidus_support/engine_extensions/decorators.rb:
```diff
@@ -4,33 +4,62 @@ module SolidusSupport
+        # won't be loaded by default. We need them to be executed regardless in order to decorate
+        # existing classes.
+        def load_decorators_from(path)
+          Dir.glob(path.join('**/*.rb')) do |decorator_path|
```
`path.glob('**/*.rb')`

kennyadsl reviewed (commented):
Super! Just left a couple of things.

spaghetticode commented:
@aldesantis I have one concern. 

We're adding new private methods to a class that inherits form `Rails::Engine`, a class that we don't own. I wonder if the new methods, having a rather generic name, may end up in overriding the ones that could be defined by the superclass in a remote but still plausible future. 

So, instead of for example `decorators_root`, would something like `solidus_decorators_root` make more sense?

aldesantis commented:
@spaghetticode I think that's a good point. It seems unlikely that Rails will officially introduce the concept of engines at some point, but it doesn't hurt to give the methods names that are a bit more specific.

spaghetticode reviewed (approved):
@aldesantis thank you 👍

aldesantis committed:
Move EngineExtensions::Decorators to EngineExtensions

This module doesn't just deal with decorators anymore, but also
loads views and controllers, so this seems like a better name.
Separating it into two modules wouldn't be simple because they
are interlinked.

kennyadsl reviewed (approved):
C O O L!
